### PR TITLE
Lock initialization, handlers and core logic

### PR DESCRIPTION
**Pull request by dexX7**
... as well as the pending transactions and global balance cache.

This should do the trick for the core, but there are still unguarded reads from the UI and RPC for globals other than `my_tally_map`, `my_pending` or the balance cache.

### CODE REVIEW

**Comment by zathras-crypto** [comment]
Reviewed, looks good to me - thanks bud :) merge at your convenience

**Comment by dexX7** [comment]
Thanks. :)

I was wondering about the wallet transaction cache. Is it going to be used?

**Comment by zathras-crypto** [comment]
> I was wondering about the wallet transaction cache. Is it going to be used?

Well, I was working on some optimization and that's where the tx hash cache came in - but it's effectiveness is limited to the scenario where we have to search through lots of bitcoin transactions to find our Omni transactions...  Eg, let's assume an integrator (say an exchange) is using Omni Core for both their Bitcoin and Omni handling (a risk at this stage but let's assume):

* Their wallet has 10K bitcoin transactions in it
* Their wallet has 1K Omni transactions in it
* The transactions are interspersed fairly randomly

In this scenario the wallet tx hash cache can really help as it avoids the need to check non-Omni transactions - *but* - with the work you've done on serialization and input caching you've already made much greater strides on performance than the wallet tx hash cache can ever provide, and since it's effectiveness is limited to that one scenario - it's a bit of a hmm...

So in a nutshell, I'm not sure whether it'll actually get used :) - what do you think?

**Comment by dexX7** [comment]
Ah, I was just wondering, because I didn't find any references to the cache. :)

> what do you think?

Let's wait and see. I have not really an idea how it could be used, but I also don't see a reason to remove it right now.

**Comment by zathras-crypto** [comment]
> Ah, I was just wondering, because I didn't find any references to the cache. :)

Ahh - yeah I removed them in https://github.com/zathras-crypto/omnicore/commit/bec0f226974f41eb6814c9295ae6e43f94038f44 and https://github.com/zathras-crypto/omnicore/commit/7183cf9962693bd489bc0f0bc72d1668ab1ade55 :)
